### PR TITLE
Add dimension, metric and namespace validation

### PR DESCRIPTION
**Pull request by markkuhn**
Related to:
* https://github.com/awslabs/aws-embedded-metrics-node/pull/135
* https://github.com/awslabs/aws-embedded-metrics-java/pull/119
* https://github.com/awslabs/aws-embedded-metrics-python/pull/91

*Description of changes:*
Added validation for dimension, metric and namespace

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/Amazon.CloudWatch.EMF.Tests/Model/DimensionSetTests.cs [hunk 1]
```diff
@@ -25,6 +26,43 @@ public void AddDimension_30_Dimensions()
...
+            });
+        }
+
+        [Fact]
+        public void AddDimension_WithNameTooLong_ThrowsInvalidDimensionException()
+        {
+            Assert.Throws<InvalidDimensionException>(() =>
+            {
+                var dimensionSet = new DimensionSet();
+                dimensionSet.AddDimension(new string('a', Constants.MAX_DIMENSION_NAME_LENGTH + 1), "value");
```
Comment (amazon-ankalini) [inline tests/Amazon.CloudWatch.EMF.Tests/Model/DimensionSetTests.cs:52]: @markkuhn probably will be more readable if you move out the string construction into a separate variable like in another tests

**Review by amazon-ankalini — CHANGES_REQUESTED** [review]

File: src/Amazon.CloudWatch.EMF/Utils/Validator.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+using System.Text;
+using Amazon.CloudWatch.EMF.Model;
+
+namespace Amazon.CloudWatch.EMF.Utils
+{
+    public class Validator
+    {
+        internal static void ValidateDimensionSet(string dimensionName, string dimensionValue)
+        {
+            if (dimensionName == null || dimensionName.Trim().Length == 0)
```
[the remark this answers is not in the document]
Comment (markkuhn) [reply src/Amazon.CloudWatch.EMF/Utils/Validator.cs:11]: In what package can I find `isBlank()`?

File: src/Amazon.CloudWatch.EMF/Utils/Validator.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Text;
+using Amazon.CloudWatch.EMF.Model;
+
+namespace Amazon.CloudWatch.EMF.Utils
+{
+    public class Validator
+    {
+        internal static void ValidateDimensionSet(string dimensionName, string dimensionValue)
```
[the remark this answers is not in the document]
Comment (markkuhn) [reply src/Amazon.CloudWatch.EMF/Utils/Validator.cs:9]: Namespaces cannot contains methods. They can only contains nested namespaces, structs and classes.

[the remark this answers is not in the document]
Comment (markkuhn) [reply src/Amazon.CloudWatch.EMF/Utils/Validator.cs:9]: This is not the case in C#

File: src/Amazon.CloudWatch.EMF/Model/DimensionSet.cs [hunk 4]
```diff
@@ -19,6 +20,7 @@ public DimensionSet()
         /// <param name="value">the value for the dimension</param>
         public DimensionSet(string key, string value)
         {
+            Validator.ValidateDimensionSet(key, value);
```
[the remark this answers is not in the document]
Comment (markkuhn) [reply src/Amazon.CloudWatch.EMF/Model/DimensionSet.cs:23]: According to [this](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/constructor) page, it is appropriate to throw constructors in C#.

File: src/Amazon.CloudWatch.EMF/Utils/Validator.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                throw new InvalidMetricException($"Metric name {name} cannot be longer than {Constants.MAX_METRIC_NAME_LENGTH} characters");
+            }
+
+            if (!Double.IsFinite(value))
+            {
+                throw new InvalidMetricException($"Metric value {value} must be a finite number");
+            }
+        }
+
+        internal static void ValidateNamespace(string namespaceName)
```
[the remark this answers is not in the document]
Comment (markkuhn) [reply src/Amazon.CloudWatch.EMF/Utils/Validator.cs:65]: I've marked them as `readonly` with the `in` keyword.

File: src/Amazon.CloudWatch.EMF/Utils/Validator.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (amazon-ankalini) [reply src/Amazon.CloudWatch.EMF/Utils/Validator.cs:10]: @markkuhn in Java package :) Nevermind that

**Review by amazon-ankalini — APPROVED** [review]

**Review by Himtanaya — APPROVED** [review]

**Pull request merged**